### PR TITLE
Initial pg package & deploy on version dump

### PR DESCRIPTION
**Pull request by georg-schwarz**

### CODE REVIEW

File: .github/workflows/basics.yml [hunk 1]
```diff
@@ -1,22 +1,33 @@
...
+name: Build and publish basics lib
 
-on:
-  release:
-    types: [published]
+on: [push, pull_request, workflow_dispatch]
 
 jobs:
-  build-and-publish:
+  pg-build:
```
Comment (mathiaszinnen) [inline .github/workflows/basics.yml]: why pg? shouldn't it be basics here?

File: .github/workflows/pg.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+name: Build and publish pg lib
```
Comment (mathiaszinnen) [inline .github/workflows/pg.yml]: I would prefer the unabbreviated "postgres lib" in the workflow name. It is easier to grasp what's going on then

File: basics/package.json [hunk 3]
```diff
@@ -11,7 +11,9 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "lint": "eslint src --fix",
     "lint-ci": "eslint src --max-warnings=0",
-    "build": "tsc"
+    "build": "tsc",
+    "get:name": "echo $npm_package_name",
```
Comment (mathiaszinnen) [inline basics/package.json:15]: I have never tried it, but you could use [npm-version](https://docs.npmjs.com/cli/v6/commands/npm-version) here, too. Maybe this simplifies some of the work you have to do in the script.

File: README.md [hunk 4]
```diff
@@ -14,7 +14,8 @@ Add the packages to your `package.json` as usual or use `npm install --save <pac
 
 | **Name** | **Package Name** | **Current Version** | **Description** |
 |---|---|---|---|
-| Basics | `@jvalue/node-dry-basics` | `0.0.2` | *all basic functionality that need no dependencies* |
+| Basics | `@jvalue/node-dry-basics` | `0.0.3` | *all basic functionality that need no dependencies* |
+| Postgres | `@jvalue/node-dry-pg` | `0.0.1` | *postgres helpers and the pg package* |
```
Comment (mathiaszinnen) [inline README.md]: Pg's package.json says it is on version 0.0.2. 

I don't know how much effort is needed to achieve that, but using npm-version we could maybe automatically update the version numbers in the README file

File: pg/package.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@jvalue/node-dry-pg",
+  "version": "0.0.2",
+  "description": "Reusable pg helpers for our node projects (NOT containing any dependencies)",
```
Comment (mathiaszinnen) [inline pg/package.json]: What about pg and dry-basics dependencies?

File: basics/package.json (hunk 3 above)
Comment (georg-schwarz) [reply basics/package.json:15]: `npm version` just manages versions. However, I only want to deploy when we adjust the version in the package.json... So you can use `npm version` to bump to a new patch version e.g., but still you have to deploy. So this is more complementary.

**Review by andreas-bauer — COMMENTED** [review]
Here are my 2 cents to this PR, but more about its repository.

### Multiple projects in one repo
In my opinion the current constellation of multiple projects under one repository causes unnecessary complexity (see publish script) and is not what I would expect from a library repository. 

### publish_on_version_bump.sh
Why is the publish process that complicated that it requires a custom 46 line bash script?
My expectation is that the CI workflow will publish the current build if the commit has a git tag. If this repo serves as a library it should be tagged properly anyways.  

### Provided functionalities 
While the `pg` project provides a useful db helper, which would be potentially not implemented well in a new service, the functionalities of the `basics` projects are doubtfully. 

`readEnvOrDie` in `env.ts`: I don't think this is a good solution where a lib can lead the main application to exit, even it is the main purpose of the function. The function should highlight problems and the main application can decide how to handle this exception.

`sleep.ts`, `validators`: The simple functionalities of these files like `sleep()` or `isString()` are not worth creating a dependency to an external lib, IMHO. 

### My recommendation
(A) Split up the different projects into separate repositories. This way you could tag every repository with proper Git tags that reflect the version of the project. Each repository could than also have a badge with the current version its project.

(B) Create templates instead of libs for common functionalities that can be slightly adjusted if needed. E.g. not all validators of a validator template are needed, or a special one is required only for one service. Keep this templates in the mono repo.

I understand that you are trying to avoid that the services drift apart in their implementation and avoid unnecessary re-implementation of already solved problems.
But I have the fear that this repository will not help you to increase the productivity during the development of the main project. I did not liked the workflow that came with the `jvalue/commons` module in the ODS v1 project.  

A small fix becomes unnecessary complicated, especially if it needs to be tested in the main application.

1. Code change in the lib
2. Publish lib with new version
3. run npm install in main application (also clean npm cache if version number didn't changed)
4. Test new lib in main application
5. Lib does not work as expected, or results in a conflict with libs or enviroments of the main application.
6. Refactor lib 
6. Publish lib with new version
8. Repeat step 3. until it works in every sub-project/service.

Sorry for the rant :v:

**Comment by georg-schwarz** [comment]
Thanks for your feedback, I appreciate it very much!

I agree with your point that the multi-lib-in-one-repo approach introduces too much complexity (see the bash script). Probably we should change that. It feels kind of strange to have a repository with one file (in the pg example), but whatever.

The thing that caused us problems in the jvalue/commons module was the fact that business logic was included there. But I hear you that e.g. the validators should not be there since it is microservice specific what you want to check.

I suggest that we discuss this further together in the next developer meeting!

**Comment by MBuchalik** [comment]
> readEnvOrDie in env.ts: I don't think this is a good solution where a lib can lead the main application to exit, even it is the main purpose of the function. The function should highlight problems and the main application can decide how to handle this exception.

The idea behind this library is to reduce the need for duplicate code between the microservices, am I right? So if you see that, in every microservice, you need a function that kills the process if some condition is not met, then you have two options:
- Declare such a function in every microservice.
- Declare such a function in a library and give it a name that indicates the potential "danger".

So, I don't really see a big problem here. The only thing you could really argue is the absence of a "readEnv" function that maybe returns `undefined` (or throws an Exception) when something went wrong.



> A small fix becomes unnecessary complicated, especially if it needs to be tested in the main application.

How about seeing it the other way round? Let's say you have created a shared library and put it directly into the monorepo. Now, every time you change something in this library, you **must** test every single microservice with this change. If you have lots of microservices, then this task becomes pretty time-consuming. And if this was a breaking change, you are now forced to update the code of every microservice at once.

If you have a versioned npm package, then every microservice can be updated independently. So, if you introduce a breaking change to the shared package, then you can gradually update the microservices (one-by-one).

Isn't that rather the idea behind a microservice?

> Split up the different projects into separate repositories.

Makes sense.

> Create templates instead of libs for common functionalities that can be slightly adjusted if needed.

Not sure what a "template" is. Will maybe ask Georg about that today in the meeting.


> sleep.ts, validators: The simple functionalities of these files like sleep() or isString() are not worth creating a dependency to an external lib, IMHO.

The question is: Are there more validators or util functions that we could put into this repo? If no, then you are right. But if we find that, over time, this repo would indeed grow, then the repo will gain more and more value.


--------

I think one of the most important questions we need to ask is: Should microservices be independent from each other, or not? Or, in other words: Should it be possible to move one microservice out of a monorepo into a separate one? 
Once we start putting shared libs into the monorepo, then the microservices become more "glued together". Is that what we want, or do we rather want them to stay independent? (Minus the communication API between them - this is probably another topic.)

**Comment by georg-schwarz** [comment]
=> will be moved to own repository. Closing...

**Pull request closed without merging**